rasa/nlu/classifiers/diet_classifier.py

```diff
@@ -0,0 +1,1424 @@
+                sparse_features.append(_sparse)
+            if _dense is not None:
+                dense_features.append(_dense)
+
+        sparse_features = np.array(sparse_features)
+        dense_features = np.array(dense_features)
+
+        return [sparse_features, dense_features]
+
+    @staticmethod
+    def _compute_default_label_features(
+        labels_example: List[Message],
+    ) -> List[np.ndarray]:
+        """Compute one-hot representation for the labels"""
+
+        return [
+            np.array(
+                [
+                    np.expand_dims(a, 0)
+                    for a in np.eye(len(labels_example), dtype=np.float32)
+                ]
+            )
+        ]
+
+    def _create_label_data(
+        self,
+        training_data: TrainingData,
+        label_id_dict: Dict[Text, int],
+        attribute: Text,
+    ) -> RasaModelData:
+        """Create matrix with label_ids encoded in rows as bag of words.
+
+        Find a training example for each label and get the encoded features
+        from the corresponding Message object.
+        If the features are already computed, fetch them from the message object
+        else compute a one hot encoding for the label as the feature vector.
+        """
+
+        # Collect one example for each label
+        labels_idx_example = []
+        for label_name, idx in label_id_dict.items():
+            label_example = self._find_example_for_label(
+                label_name, training_data.intent_examples, attribute
+            )
+            labels_idx_example.append((idx, label_example))
+
+        # Sort the list of tuples based on label_idx
+        labels_idx_example = sorted(labels_idx_example, key=lambda x: x[0])
+        labels_example = [example for (_, example) in labels_idx_example]
+
+        # Collect features, precomputed if they exist, else compute on the fly
+        if self._check_labels_features_exist(labels_example, attribute):
+            features = self._extract_labels_precomputed_features(
+                labels_example, attribute
+            )
+        else:
+            features = self._compute_default_label_features(labels_example)
+
+        label_data = RasaModelData()
+        label_data.add_features("label_features", features)
+
+        label_ids = np.array([idx for (idx, _) in labels_idx_example])
+        # explicitly add last dimension to label_ids
+        # to track correctly dynamic sequences
+        label_data.add_features("label_ids", [np.expand_dims(label_ids, -1)])
+
+        label_data.add_mask("label_mask", "label_features")
+
+        return label_data
+
+    def _use_default_label_features(self, label_ids: np.ndarray) -> List[np.ndarray]:
+        return [
+            np.array(
+                [
+                    self._label_data.get("label_features")[0][label_id]
+                    for label_id in label_ids
+                ]
+            )
+        ]
+
+    def _create_model_data(
```

Comment:
but it is linear, not sure helper functions will make it clearer

Reply:
note that this optional.
For me splitting this up is not because this is not linear, but to abstract away the details of the operations, so that readers can easily decide which parts are relevant to them and then dive into these parts